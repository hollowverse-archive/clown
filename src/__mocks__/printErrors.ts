export const printErrors = jest.fn();
