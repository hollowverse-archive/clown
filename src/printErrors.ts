import diff from 'jest-diff';
import { Discrepancies } from './types';
import _ from 'lodash';
import chalk from 'chalk';

// tslint:disable:no-console
export function printErrors(discrepancies: Discrepancies) {
  _.each(discrepancies, ({ expected, received }, filePath) => {
    console.log(`\n\n${chalk.bold.underline(filePath)}`);

    console.log(
      diff(expected, received, {
        aAnnotation: 'What it should be',
        bAnnotation: 'What it is',
      }),
    );
  });
}
