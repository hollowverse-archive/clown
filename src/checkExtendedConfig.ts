import { readFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import { FileContents, Discrepancies } from './types';
import { printErrors } from './printErrors';
import { isJsonEqual } from './isJsonEqual';

export async function checkExtendedConfig(cwd: string) {
  const expectedContents = await computeFileContents(cwd);
  const iterableExpectedContents = _.entries(expectedContents);
  const currentContent: FileContents = await bluebird.reduce(
    iterableExpectedContents,
    async (currentContentAccumulator: FileContents, [destinationPath]) => {
      const content = await readFile(destinationPath);

      if (content) {
        currentContentAccumulator[destinationPath] = content;
      }

      return currentContentAccumulator;
    },
    {},
  );

  const discrepancies: Discrepancies = await bluebird.reduce(
    iterableExpectedContents,
    async (
      discrepancyAccumulator: Discrepancies,
      [destinationPath, expectedContent],
    ) => {
      const destinationContent = currentContent[destinationPath];

      if (
        (destinationContent === undefined ||
          destinationContent !== expectedContent) &&
        !isJsonEqual(destinationContent, expectedContent)
      ) {
        discrepancyAccumulator[destinationPath] = {
          expected: expectedContent,
          received: destinationContent,
        };
      }

      return discrepancyAccumulator;
    },
    {},
  );

  if (_.isEmpty(discrepancies)) {
    process.exit(0);
  } else {
    printErrors(discrepancies);
    process.exit(1);
  }
}
