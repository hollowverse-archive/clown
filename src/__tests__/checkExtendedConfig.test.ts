import { checkExtendedConfig } from '../checkExtendedConfig';
import { vol } from '@forabi/memfs';
import { printErrors } from '../printErrors';

describe('checkExtendedConfig basics', () => {
  beforeEach(() => {
    vol.reset();
    // @ts-ignore
    printErrors.mockReset();
  });

  it('use case 1', async () => {
    const files = {
      '/clown.json': `{
        "extensions": ["/clownExtensionA"]
      }`,
      '/clownExtensionA/package.json': `{
        "foo": "bar"
      }`,
      '/clownExtensionA/eslintrc.json': `{
        "rules": "ok"
      }`,
      '/package.json': `{
        "name": "Stuff"
      }`,
      '/eslintrc.json': `{
        "extends": "someFoo"
      }`,
    };

    vol.fromJSON(files);

    await checkExtendedConfig('/');

    // @ts-ignore
    expect(printErrors.mock.calls).toMatchSnapshot();
  });

  it('does not complain when JSONs are equal but formatted differently', async () => {
    const files = {
      '/clown.json': `{
        "extensions": ["/clownExtensionA"]
      }`,
      '/clownExtensionA/package.json': `{
        "name": [
          "Stuff"
        ]
      }`,
      '/package.json': `{
        "name": ["Stuff"]
      }`,
    };

    vol.fromJSON(files);

    await checkExtendedConfig('/');

    // @ts-ignore
    expect(printErrors.mock.calls).toMatchSnapshot();
  });
});
