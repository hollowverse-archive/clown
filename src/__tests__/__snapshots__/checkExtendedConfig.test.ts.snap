// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkExtendedConfig basics does not complain when JSONs are equal but formatted differently 1`] = `Array []`;

exports[`checkExtendedConfig basics use case 1 1`] = `
Array [
  Array [
    Object {
      "/eslintrc.json": Object {
        "expected": "{
        \\"extends\\": \\"someFoo\\",
        \\"rules\\": \\"ok\\"
}
",
        "received": "{
        \\"extends\\": \\"someFoo\\"
      }",
      },
      "/package.json": Object {
        "expected": "{
        \\"name\\": \\"Stuff\\",
        \\"foo\\": \\"bar\\"
}
",
        "received": "{
        \\"name\\": \\"Stuff\\"
      }",
      },
    },
  ],
]
`;
